File: libraries/typescript/src/tts/elevenlabs-ws.ts
/** ElevenLabs WebSocket TTS for Patter pipeline mode (opt-in low-latency). */
import {
  ElevenLabsWebSocketTTS as _ElevenLabsWebSocketTTS,
  type ElevenLabsWebSocketTTSOptions,
} from '../providers/elevenlabs-ws-tts';
import type { ElevenLabsModel } from '../providers/elevenlabs-tts';

export type { ElevenLabsModel };

/** Constructor options for the ElevenLabs WebSocket `TTS` adapter. */
export interface ElevenLabsWebSocketOptions {
  /** API key. Falls back to ELEVENLABS_API_KEY env var when omitted. */
  apiKey?: string;
  voiceId?: string;
  modelId?: ElevenLabsModel | string;
  outputFormat?: string;
  /** Let the server pick chunk timing. Default true. */
  autoMode?: boolean;
  voiceSettings?: Record<string, unknown>;
  languageCode?: string;
  /** WS keep-alive timeout in seconds (5–180). Default 60. */
  inactivityTimeout?: number;
  /** Manual chunk schedule, only used when ``autoMode: false``. */
  chunkLengthSchedule?: number[];
}

/** Options for the carrier-specific factories — same as the constructor minus `outputFormat`. */
export type ElevenLabsWebSocketCarrierOptions = Omit<ElevenLabsWebSocketOptions, 'outputFormat'>;

function resolveApiKey(apiKey: string | undefined): string {
  const key = apiKey ?? process.env.ELEVENLABS_API_KEY;
  if (!key) {
    throw new Error(
      "ElevenLabs WebSocket TTS requires an apiKey. Pass { apiKey: '...' } or " +
        'set ELEVENLABS_API_KEY in the environment.',
    );
  }
  return key;
}

function buildOpts(opts: ElevenLabsWebSocketOptions): ElevenLabsWebSocketTTSOptions {
  // Voice ID default is owned by the provider class — passing ``undefined``
  // here lets it apply its own default (parity Python ↔ TS).
  const out: ElevenLabsWebSocketTTSOptions = {
    apiKey: resolveApiKey(opts.apiKey),
    modelId: opts.modelId ?? 'eleven_flash_v2_5',
    outputFormat: opts.outputFormat ?? 'pcm_16000',
    autoMode: opts.autoMode ?? true,
  };
  if (opts.voiceId !== undefined) out.voiceId = opts.voiceId;
  if (opts.voiceSettings !== undefined) out.voiceSettings = opts.voiceSettings;
  if (opts.languageCode !== undefined) out.languageCode = opts.languageCode;
  if (opts.inactivityTimeout !== undefined) out.inactivityTimeout = opts.inactivityTimeout;
  if (opts.chunkLengthSchedule !== undefined) out.chunkLengthSchedule = opts.chunkLengthSchedule;
  return out;
}

/**
 * ElevenLabs streaming TTS over WebSocket.
 *
 * Drop-in replacement for `getpatter/tts/elevenlabs.TTS` (HTTP) using the
 * `stream-input` WebSocket endpoint. Saves the per-utterance HTTP request
 * setup time; otherwise behaves identically.
 *
 * @example
 * ```ts
 * import * as elevenlabsWs from "getpatter/tts/elevenlabs-ws";
 * const tts = new elevenlabsWs.TTS();              // reads ELEVENLABS_API_KEY
 * const tts = elevenlabsWs.TTS.forTwilio({ apiKey: "..." });
 * ```
 *
 * **Telephony optimisation** — use {@link TTS.forTwilio} (μ-law @ 8 kHz)
 * or {@link TTS.forTelnyx} (PCM @ 16 kHz) on phone calls.
 */
export class TTS extends _ElevenLabsWebSocketTTS {
  static readonly providerKey = 'elevenlabs_ws';

  constructor(opts: ElevenLabsWebSocketOptions = {}) {
    super(buildOpts(opts));
  }

  /** WebSocket TTS pre-configured for Twilio Media Streams (`ulaw_8000`). */
  static override forTwilio(opts: ElevenLabsWebSocketCarrierOptions = {}): TTS {
    return new TTS({ ...opts, outputFormat: 'ulaw_8000' });
  }

  /** WebSocket TTS pre-configured for Telnyx (`pcm_16000`). */
  static override forTelnyx(opts: ElevenLabsWebSocketCarrierOptions = {}): TTS {
    return new TTS({ ...opts, outputFormat: 'pcm_16000' });
  }
}

File: libraries/python/getpatter/tts/elevenlabs_ws.py
"""ElevenLabs WebSocket TTS for Patter pipeline mode (opt-in low-latency)."""

from __future__ import annotations

import os
from typing import ClassVar

from getpatter.providers.elevenlabs_ws_tts import (
    ElevenLabsWebSocketTTS as _ElevenLabsWebSocketTTS,
)

__all__ = ["TTS"]


def _resolve_api_key(api_key: str | None) -> str:
    key = api_key or os.environ.get("ELEVENLABS_API_KEY")
    if not key:
        raise ValueError(
            "ElevenLabs WebSocket TTS requires an api_key. Pass api_key='...' "
            "or set ELEVENLABS_API_KEY in the environment."
        )
    return key


class TTS(_ElevenLabsWebSocketTTS):
    """ElevenLabs streaming TTS over WebSocket (``stream-input`` endpoint).

    Drop-in replacement for :class:`getpatter.tts.elevenlabs.TTS` (HTTP)
    that uses the WebSocket transport. Saves the per-utterance HTTP request
    setup time; otherwise behaves identically.

    Example::

        from getpatter.tts import elevenlabs_ws

        tts = elevenlabs_ws.TTS()              # reads ELEVENLABS_API_KEY
        tts = elevenlabs_ws.TTS(api_key="...", voice_id="EXAVITQu4vr4xnSDxMaL")

    Telephony optimization
    ----------------------
    Use :meth:`for_twilio` (μ-law @ 8 kHz, native Twilio Media Streams
    format) or :meth:`for_telnyx` (PCM @ 16 kHz) — same wire-format
    optimisation as the HTTP variant.
    """

    provider_key: ClassVar[str] = "elevenlabs_ws"

    def __init__(
        self,
        api_key: str | None = None,
        *,
        voice_id: str | None = None,
        model_id: str = "eleven_flash_v2_5",
        output_format: str = "pcm_16000",
        auto_mode: bool = True,
        voice_settings: dict | None = None,
        language_code: str | None = None,
        inactivity_timeout: int | None = None,
        chunk_length_schedule: list[int] | None = None,
    ) -> None:
        # ``voice_id`` defaults are owned by the provider class so the public
        # wrapper and the low-level class agree on the default voice. Pass
        # only when the caller specifies one.
        kwargs: dict = {
            "api_key": _resolve_api_key(api_key),
            "model_id": model_id,
            "output_format": output_format,
            "auto_mode": auto_mode,
        }
        if voice_id is not None:
            kwargs["voice_id"] = voice_id
        if voice_settings is not None:
            kwargs["voice_settings"] = voice_settings
        if language_code is not None:
            kwargs["language_code"] = language_code
        if inactivity_timeout is not None:
            kwargs["inactivity_timeout"] = inactivity_timeout
        if chunk_length_schedule is not None:
            kwargs["chunk_length_schedule"] = chunk_length_schedule
        super().__init__(**kwargs)

    @classmethod
    def for_twilio(
        cls,
        api_key: str | None = None,
        *,
        voice_id: str | None = None,
        model_id: str = "eleven_flash_v2_5",
        auto_mode: bool = True,
        voice_settings: dict | None = None,
        language_code: str | None = None,
        inactivity_timeout: int | None = None,
    ) -> "TTS":
        """WebSocket TTS pre-configured for Twilio Media Streams (``ulaw_8000``)."""
        return cls(
            api_key=api_key,
            voice_id=voice_id,
            model_id=model_id,
            output_format="ulaw_8000",
            auto_mode=auto_mode,
            voice_settings=voice_settings,
            language_code=language_code,
            inactivity_timeout=inactivity_timeout,
        )

    @classmethod
    def for_telnyx(
        cls,
        api_key: str | None = None,
        *,
        voice_id: str | None = None,
        model_id: str = "eleven_flash_v2_5",
        auto_mode: bool = True,
        voice_settings: dict | None = None,
        language_code: str | None = None,
        inactivity_timeout: int | None = None,
    ) -> "TTS":
        """WebSocket TTS pre-configured for Telnyx (``pcm_16000``)."""
        return cls(
            api_key=api_key,
            voice_id=voice_id,
            model_id=model_id,
            output_format="pcm_16000",
            auto_mode=auto_mode,
            voice_settings=voice_settings,
            language_code=language_code,
            inactivity_timeout=inactivity_timeout,
        )
